assign of an interface struct field from func call

### The following program `sample.go` triggers an unexpected result

```go
package main                                                                    
                                                                                
type T struct {         
    v interface{}
}

func f() (ret int64, err error) {
    ret += 2
    return
}

func main() {
    t := &T{}
    t.v, _ = f()
    println(t.v.(int64))
}
```


### Expected result

```console
$ go run ./sample.go
2
```


### Got

```console
$ yaegi ./sample.go
./sample.go:8:2: panic
./sample.go:13:8: panic
run: reflect: call of reflect.Value.Int on interface Value
goroutine 1 [running]:
runtime/debug.Stack()
	/Users/marc/sdk/go1.17.2/src/runtime/debug/stack.go:24 +0x88
github.com/traefik/yaegi/interp.(*Interpreter).Execute.func1(0x140001bbae0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/program.go:131 +0x8c
panic({0x102f73c00, 0x14000136690})
	/Users/marc/sdk/go1.17.2/src/runtime/panic.go:1038 +0x21c
github.com/traefik/yaegi/interp.runCfg.func1(0x14000368160, 0x140004bd0e0, 0x0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:193 +0x164
panic({0x102f73c00, 0x14000136690})
	/Users/marc/sdk/go1.17.2/src/runtime/panic.go:1038 +0x21c
github.com/traefik/yaegi/interp.runCfg.func1(0x14000368210, 0x140004bc240, 0x0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:193 +0x164
panic({0x102f73c00, 0x14000136690})
	/Users/marc/sdk/go1.17.2/src/runtime/panic.go:1038 +0x21c
reflect.Value.Int(...)
	/Users/marc/sdk/go1.17.2/src/reflect/value.go:1334
github.com/traefik/yaegi/interp.genValueInt.func1(0x14000368210)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/value.go:522 +0xfc
github.com/traefik/yaegi/interp.addAssign.func2(0x14000368210)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/op.go:1518 +0x5c
github.com/traefik/yaegi/interp.runCfg(0x140004bc240, 0x14000368210, 0x14000371320, 0x140004bd8c0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:201 +0x268
github.com/traefik/yaegi/interp.call.func6(0x14000368160)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:1343 +0x7c0
github.com/traefik/yaegi/interp.runCfg(0x140004bd0e0, 0x14000368160, 0x140004bc6c0, 0x0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:201 +0x268
github.com/traefik/yaegi/interp.(*Interpreter).run(0x14000366000, 0x140004bc6c0, 0x14000368000)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/run.go:120 +0x3d4
github.com/traefik/yaegi/interp.(*Interpreter).Execute(0x14000366000, 0x14000467650)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/program.go:157 +0x1e0
github.com/traefik/yaegi/interp.(*Interpreter).eval(0x14000366000, {0x140002de630, 0xa6}, {0x16d937710, 0x1e}, 0x0)
	/Users/marc/go/src/github.com/traefik/yaegi/interp/interp.go:615 +0x98
github.com/traefik/yaegi/interp.(*Interpreter).EvalPath(0x14000366000, {0x16d937710, 0x1e})
	/Users/marc/go/src/github.com/traefik/yaegi/interp/interp.go:508 +0xe8
main.runFile(0x14000366000, {0x16d937710, 0x1e}, 0x0)
	/Users/marc/go/src/github.com/traefik/yaegi/cmd/yaegi/run.go:154 +0xb0
main.run({0x14000032050, 0x1, 0x1})
	/Users/marc/go/src/github.com/traefik/yaegi/cmd/yaegi/run.go:117 +0xe14
main.main()
	/Users/marc/go/src/github.com/traefik/yaegi/cmd/yaegi/yaegi.go:144 +0x384
```


### Yaegi Version

v0.11.0

### Additional Notes

This issue is related to #1275

--- a/_test/issue-1311.go
+++ b/_test/issue-1311.go
@@ -1,0 +1,19 @@
+package main
+
+type T struct {
+	v interface{}
+}
+
+func f() (ret int64, err error) {
+	ret += 2
+	return
+}
+
+func main() {
+	t := &T{}
+	t.v, _ = f()
+	println(t.v.(int64))
+}
+
+// Output:
+// 2

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -64,6 +64,9 @@ func (interp *Interpreter) cfg(root *node, sc *scope, importPath, pkgName string
 		if err != nil {
 			return false
 		}
+		if n.scope == nil {
+			n.scope = sc
+		}
 		switch n.kind {
 		case binaryExpr, unaryExpr, parenExpr:
 			if isBoolAction(n) {
@@ -662,7 +665,12 @@ func (interp *Interpreter) cfg(root *node, sc *scope, importPath, pkgName string
 				if r := lc.child[0].typ.numOut(); r != l {
 					err = n.cfgErrorf("assignment mismatch: %d variables but %s returns %d values", l, lc.child[0].name(), r)
 				}
-				n.gen = nop
+				if isBinCall(lc, sc) {
+					n.gen = nop
+				} else {
+					// TODO (marc): skip if no conversion or wrapping is needed.
+					n.gen = assignFromCall
+				}
 			case indexExpr:
 				lc.gen = getIndexMap2
 				n.gen = nop
@@ -970,7 +978,7 @@ func (interp *Interpreter) cfg(root *node, sc *scope, importPath, pkgName string
 					n.typ = c0.typ
 					n.findex = sc.add(n.typ)
 				}
-			case isBinCall(n):
+			case isBinCall(n, sc):
 				err = check.arguments(n, n.child[1:], n.child[0], n.action == aCallSlice)
 				if err != nil {
 					break
@@ -2019,7 +2027,12 @@ func compDefineX(sc *scope, n *node) error {
 		if len(types) != l {
 			return n.cfgErrorf("assignment mismatch: %d variables but %s returns %d values", l, src.child[0].name(), len(types))
 		}
-		n.gen = nop
+		if isBinCall(src, sc) {
+			n.gen = nop
+		} else {
+			// TODO (marc): skip if no conversion or wrapping is needed.
+			n.gen = assignFromCall
+		}
 
 	case indexExpr:
 		types = append(types, src.typ, sc.getType("bool"))
@@ -2468,8 +2481,19 @@ func isCall(n *node) bool {
 	return n.action == aCall || n.action == aCallSlice
 }
 
-func isBinCall(n *node) bool {
-	return isCall(n) && n.child[0].typ.cat == valueT && n.child[0].typ.rtype.Kind() == reflect.Func
+func isBinCall(n *node, sc *scope) bool {
+	if !isCall(n) || len(n.child) == 0 {
+		return false
+	}
+	c0 := n.child[0]
+	if c0.typ == nil {
+		// If called early in parsing, child type may not be known yet.
+		c0.typ, _ = nodeType(n.interp, sc, c0)
+		if c0.typ == nil {
+			return false
+		}
+	}
+	return c0.typ.cat == valueT && c0.typ.rtype.Kind() == reflect.Func
 }
 
 func isOffsetof(n *node) bool {

--- a/interp/run.go
+++ b/interp/run.go
@@ -616,6 +616,34 @@ func convert(n *node) {
 	}
 }
 
+// assignFromCall assigns values from a function call.
+func assignFromCall(n *node) {
+	ncall := n.lastChild()
+	l := len(n.child) - 1
+	if n.anc.kind == varDecl && n.child[l-1].isType(n.scope) {
+		// Ignore the type in the assignment if it is part of a variable declaration.
+		l--
+	}
+	dvalue := make([]func(*frame) reflect.Value, l)
+	for i := range dvalue {
+		if n.child[i].ident == "_" {
+			continue
+		}
+		dvalue[i] = genValue(n.child[i])
+	}
+	next := getExec(n.tnext)
+	n.exec = func(f *frame) bltn {
+		for i, v := range dvalue {
+			if v == nil {
+				continue
+			}
+			s := f.data[ncall.findex+i]
+			v(f).Set(s)
+		}
+		return next
+	}
+}
+
 func assign(n *node) {
 	next := getExec(n.tnext)
 	dvalue := make([]func(*frame) reflect.Value, n.nleft)
@@ -1126,7 +1154,7 @@ func call(n *node) {
 	// Compute input argument value functions.
 	for i, c := range child {
 		switch {
-		case isBinCall(c):
+		case isBinCall(c, c.scope):
 			// Handle nested function calls: pass returned values as arguments.
 			numOut := c.child[0].typ.rtype.NumOut()
 			for j := 0; j < numOut; j++ {
@@ -1169,6 +1197,7 @@ func call(n *node) {
 	rvalues := make([]func(*frame) reflect.Value, len(rtypes))
 	switch n.anc.kind {
 	case defineXStmt, assignXStmt:
+		l := n.level
 		for i := range rvalues {
 			c := n.anc.child[i]
 			switch {
@@ -1177,7 +1206,8 @@ func call(n *node) {
 			case isInterfaceSrc(c.typ) && !isEmptyInterface(c.typ) && !isInterfaceSrc(rtypes[i]):
 				rvalues[i] = genValueInterfaceValue(c)
 			default:
-				rvalues[i] = genValue(c)
+				j := n.findex + i
+				rvalues[i] = func(f *frame) reflect.Value { return getFrame(f, l).data[j] }
 			}
 		}
 	case returnStmt:
@@ -1404,7 +1434,7 @@ func callBin(n *node) {
 		}
 
 		switch {
-		case isBinCall(c):
+		case isBinCall(c, c.scope):
 			// Handle nested function calls: pass returned values as arguments
 			numOut := c.child[0].typ.rtype.NumOut()
 			for j := 0; j < numOut; j++ {

